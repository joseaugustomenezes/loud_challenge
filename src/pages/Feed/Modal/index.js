import React from "react";
import BootstrapModal from "react-bootstrap/Modal";
import { Formik } from "formik";
import Form from "react-bootstrap/Form";
import { useDispatch, useSelector } from "react-redux";

import LoadingButton from "../../../components/LoadingButton";
import { createOpinion } from "../../../store/actions";

const Modal = ({ visible, onClose }) => {
  const dispatch = useDispatch();
  const opinion = useSelector((state) => state.opinion);

  const createOpinionForm = () => {
    return (
      <Formik
        initialValues={{ title: "", content: "" }}
        validate={(values) => {
          const errors = {};
          if (!values.title) errors.title = "Campo obrigatório";
          if (!values.content) errors.content = "Campo obrigatório";
          return errors;
        }}
        onSubmit={(values) => {
          dispatch(createOpinion(values));
        }}
      >
        {({
          values,
          handleSubmit,
          handleChange,
          handleBlur,
          touched,
          errors,
        }) => (
          <Form onSubmit={handleSubmit}>
            <Form.Group>
              <Form.Label>Título</Form.Label>
              <Form.Control
                type="text"
                name="title"
                id="title"
                placeholder="Título"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.username}
                isInvalid={touched.title && errors.title}
              />
              <Form.Control.Feedback type="invalid">
                {errors.title}
              </Form.Control.Feedback>
            </Form.Group>
            <Form.Group>
              <Form.Label>Descrição</Form.Label>
              <Form.Control
                as="textarea"
                type="text"
                name="content"
                ogin
                id="content"
                placeholder="Descrição"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.content}
                isInvalid={touched.content && errors.content}
              />
              <Form.Control.Feedback type="invalid">
                {errors.content}
              </Form.Control.Feedback>
            </Form.Group>
            <div style={{ textAlign: "end" }}>
              <LoadingButton loading={opinion?.loading} type="submit">
                Criar
              </LoadingButton>
            </div>
          </Form>
        )}
      </Formik>
    );
  };

  return (
    <BootstrapModal show={visible} onHide={() => onClose()}>
      <BootstrapModal.Header closeButton>
        <BootstrapModal.Title>Nova opinião</BootstrapModal.Title>
      </BootstrapModal.Header>
      <BootstrapModal.Body>{createOpinionForm()}</BootstrapModal.Body>
    </BootstrapModal>
  );
};

export default Modal;
