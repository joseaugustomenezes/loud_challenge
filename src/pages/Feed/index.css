.card-text img {
  width: 100%;
}