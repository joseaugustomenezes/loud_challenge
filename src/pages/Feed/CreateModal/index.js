import React from "react";
import Modal from "react-bootstrap/Modal";
import { Formik } from "formik";
import Form from "react-bootstrap/Form";
import { useDispatch, useSelector } from "react-redux";

import LoadingButton from "../../../components/LoadingButton";
import { createOpinion } from "../../../store/actions";

const CreateModal = ({ visible, onClose }) => {
  const dispatch = useDispatch();
  const opinion = useSelector((state) => state.createOpinion);

  const MyInnerForm = ({
    values,
    handleSubmit,
    handleChange,
    handleBlur,
    touched,
    errors,
  }) => (
    <Form onSubmit={handleSubmit}>
      <Form.Group>
        <Form.Label>Título</Form.Label>
        <Form.Control
          type="text"
          name="title"
          id="title"
          placeholder="Título"
          onChange={handleChange}
          onBlur={handleBlur}
          value={values.username}
          isInvalid={touched.title && errors.title}
        />
        <Form.Control.Feedback type="invalid">
          {errors.title}
        </Form.Control.Feedback>
      </Form.Group>
      <Form.Group>
        <Form.Label>Descrição</Form.Label>
        <Form.Control
          as="textarea"
          type="text"
          name="content"
          ogin
          id="content"
          placeholder="Descrição"
          onChange={handleChange}
          onBlur={handleBlur}
          value={values.content}
          isInvalid={touched.content && errors.content}
        />
        <Form.Control.Feedback type="invalid">
          {errors.content}
        </Form.Control.Feedback>
      </Form.Group>
      <div style={{ textAlign: "end" }}>
        <LoadingButton loading={opinion?.loading} type="submit">
          Criar
        </LoadingButton>
      </div>
    </Form>
  );

  const createOpinionForm = () => {
    return (
      <Formik
        initialValues={{ title: "", content: "" }}
        validate={(values) => {
          const errors = {};
          if (!values.title) errors.title = "Campo obrigatório";
          if (!values.content) errors.content = "Campo obrigatório";
          return errors;
        }}
        onSubmit={(values) => {
          dispatch(createOpinion(values));
        }}
      >
        {(props) => <MyInnerForm {...props} />}
      </Formik>
    );
  };

  return (
    <Modal show={visible} onHide={() => onClose()}>
      <Modal.Header closeButton>
        <Modal.Title>Nova opinião</Modal.Title>
      </Modal.Header>
      <Modal.Body>{createOpinionForm()}</Modal.Body>
    </Modal>
  );
};

export default CreateModal;
