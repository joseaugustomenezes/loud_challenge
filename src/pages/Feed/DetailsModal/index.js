import React, { useEffect } from "react";
import Modal from "react-bootstrap/Modal";
import { useDispatch, useSelector } from "react-redux";
import Spinner from "react-bootstrap/Spinner";
import ReactMarkdown from "react-markdown";

import {
  insertUpvote,
  deleteUpvote,
  fetchOpinion,
} from "../../../store/actions";
import LoadingButton from '../../../components/LoadingButton';
import { getUserId } from "../../../services/auth";
import "./index.css";

const DetailsModal = ({ visible, onClose, opinionId }) => {
  const opinion = useSelector((state) => state.opinion);
  const dispatch = useDispatch();

  useEffect(() => {
    opinionId && dispatch(fetchOpinion(opinionId));
  }, [opinionId, dispatch]);

  useEffect(() => {
    opinion?.error && onClose();
  }, [opinion, onClose]);

  const Spin = () => (
    <div style={{ textAlign: "center" }}>
      <Spinner style={{ marginRight: "5px" }} animation="border" role="status">
        <span className="sr-only">Loading...</span>
      </Spinner>
    </div>
  );

  const getUpvoteButton = () => {
    const userId = getUserId();
    const hasVoted = opinion?.upvotes.find(
      (upvote) => upvote.user_id === userId
    );
    return hasVoted ? (
      <LoadingButton
        loading={opinion?.loadingUpvote}
        size="sm"
        onClick={() => dispatch(deleteUpvote(opinion?.id))}
        variant="danger"
      >
        Remover
      </LoadingButton>
    ) : (
      <LoadingButton
        loading={opinion?.loadingUpvote}
        size="sm"
        onClick={() => dispatch(insertUpvote(opinion?.id))}
        variant="success"
      >
        Votar
      </LoadingButton>
    );
  };

  return (
    <Modal show={visible} onHide={() => onClose()}>
      <Modal.Header closeButton>
        <Modal.Title>
          {opinion?.loading ? "Carregando..." : opinion?.title}
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
        {opinion?.loading ? (
          <Spin />
        ) : (
          <>
            <div
              style={{
                display: "flex",
                alignItems: "center",
                justifyContent: "flex-end",
              }}
            >
              <span
                style={{
                  marginRight: "15px",
                  color: "#28a745",
                  fontWeight: "bold",
                }}
              >
                ({opinion?.upvotes?.length || 0}) Votos
              </span>
              {getUpvoteButton()}
            </div>
            <ReactMarkdown className="markdown" source={opinion?.content} />
          </>
        )}
      </Modal.Body>
    </Modal>
  );
};

export default DetailsModal;
