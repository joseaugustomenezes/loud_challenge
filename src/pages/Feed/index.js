import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import Card from "react-bootstrap/Card";
import Container from "react-bootstrap/Container";
import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import Spinner from "react-bootstrap/Spinner";
import ReactMarkdown from "react-markdown";

import Toast from "../../components/Toast";
import CreateModal from "./CreateModal";
import DetailsModal from "./DetailsModal";
import "./index.css";

import { fetchOpinions, logout } from "../../store/actions";

const Feed = () => {
  const dispatch = useDispatch();
  const opinions = useSelector((state) => state.opinions);
  const opinion = useSelector((state) => state.createOpinion);
  const [opinionsIds, setOpinionsIds] = useState([]);
  const [search, setSearch] = useState("");
  const [createModalVisible, setCreateModalVisible] = useState(false);
  const [detailsModalVisible, setDetailsModalVisible] = useState(false);
  const [detailsOpinionId, setDetailsOpinionId] = useState();

  const ids = opinions?.ids;
  const content = opinions?.content;

  useEffect(() => {
    dispatch(fetchOpinions());
  }, [dispatch]);

  useEffect(() => {
    opinions?.ids && setOpinionsIds(opinions.ids);
  }, [opinions]);

  useEffect(() => {
    opinion?.id && setCreateModalVisible(false);
  }, [opinion]);

  useEffect(() => {
    if (ids?.length) {
      search.length
        ? setOpinionsIds(
            ids.filter((opinionId) =>
              content[opinionId].title
                .toLowerCase()
                .includes(search.toLowerCase())
            )
          )
        : setOpinionsIds(ids);
    }
  }, [search, content, ids]);

  return (
    <div style={{ position: "relative" }}>
      <Toast
        message={opinions?.error?.message}
        status={opinions?.error?.status}
      />
      <Container>
        <div
          style={{
            display: "flex",
            alignItems: "center",
            justifyContent: "space-between",
          }}
        >
          <h1>Feed de Opiniões</h1>
          <Button onClick={() => dispatch(logout())}>Log Out</Button>
        </div>
        <hr />
        <Form.Group>
          <Form.Control
            type="email"
            name="email"
            id="email"
            placeholder="Buscar pelo título"
            onChange={(e) => setSearch(e.target.value)}
          />
        </Form.Group>
        <hr />
        <div style={{ textAlign: "center", marginBottom: "15px" }}>
          <Button onClick={() => setCreateModalVisible(true)}>
            Nova opinião
          </Button>
        </div>
        {opinions?.loading ? (
          <div style={{ textAlign: "center" }}>
            <Spinner
              style={{ marginRight: "5px" }}
              animation="border"
              role="status"
            >
              <span className="sr-only">Loading...</span>
            </Spinner>
          </div>
        ) : (
          opinionsIds.map((opinionId) => (
            <Card key={opinionId} style={{ marginBottom: "15px" }}>
              <Card.Body>
                <Card.Title
                  style={{
                    display: "flex",
                    alignItems: "center",
                    justifyContent: "space-between",
                  }}
                >
                  <div>{opinions.content[opinionId].title}</div>
                  <div
                    style={{
                      color: "#28a745",
                      fontWeight: "bold",
                    }}
                  >
                    ({opinions.content[opinionId].upvotes_count}) Votos
                  </div>
                </Card.Title>
                <Card.Text>
                  <div className="opinion-description">
                    <div className="shadow"></div>
                    <ReactMarkdown
                      source={opinions.content[opinionId].content}
                    />
                  </div>
                  <Button
                    style={{ marginTop: "10px" }}
                    variant="link"
                    onClick={() => {
                      setDetailsModalVisible(true);
                      setDetailsOpinionId(opinionId);
                    }}
                  >
                    Exibir Detalhes
                  </Button>
                </Card.Text>
              </Card.Body>
            </Card>
          ))
        )}
      </Container>
      <CreateModal
        visible={createModalVisible}
        onClose={() => setCreateModalVisible(false)}
      />
      <DetailsModal
        visible={detailsModalVisible}
        onClose={() => setDetailsModalVisible(false)}
        opinionId={detailsOpinionId}
      />
    </div>
  );
};

export default Feed;
