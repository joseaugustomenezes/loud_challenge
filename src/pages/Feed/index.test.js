import React from "react";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import Feed from "./index";

describe("<Feed />", () => {
  it("should render a card", () => {
    const mockStore = configureStore();
    const component = mount(
      <Provider
        store={mockStore({
          opinions: {
            ids: [0],
            content: {
              0: {
                content: "test_content",
                title: "test_title",
                upvotes_count: 3,
              },
            },
          },
        })}
      >
        <Feed />
      </Provider>
    );
    expect(component.find(".card")).toHaveLength(1);
    expect(component).toMatchSnapshot();
    component.unmount();
  });

  it("should open create modal on button click", () => {
    const mockStore = configureStore();
    const component = mount(
      <Provider store={mockStore({})}><Feed /></Provider>
    );
    component.find('.btn-primary').at(1).simulate('click');
    expect(component.find('CreateModal').prop('visible')).toBeTruthy();
  });

  it("should open details modal on button click", () => {
    const mockStore = configureStore();
    const component = mount(
      <Provider
        store={mockStore({
          opinions: {
            ids: [0],
            content: {
              0: {
                content: "test_content",
                title: "test_title",
                upvotes_count: 3,
              },
            },
          },
        })}
      >
        <Feed />
      </Provider>
    );
    component.find('.btn-link').simulate('click');
    expect(component.find('DetailsModal').prop('visible')).toBeTruthy();
  });
});
