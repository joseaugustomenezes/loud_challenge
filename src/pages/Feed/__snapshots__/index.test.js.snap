// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Feed /> should render a card 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Feed>
    <div
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <Toast>
        <div
          style={
            Object {
              "position": "absolute",
              "right": 20,
              "top": 10,
            }
          }
        >
          <Toast
            autohide={true}
            delay={3000}
            onClose={[Function]}
            show={false}
          >
            <Fade
              appear={false}
              in={false}
              mountOnEnter={false}
              timeout={300}
              unmountOnExit={true}
            >
              <Transition
                addEndListener={[Function]}
                appear={false}
                enter={true}
                exit={true}
                in={false}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={300}
                unmountOnExit={true}
              />
            </Fade>
          </Toast>
        </div>
      </Toast>
      <Container
        fluid={false}
      >
        <div
          className="container"
        >
          <div
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "justifyContent": "space-between",
              }
            }
          >
            <h1>
              Feed de Opiniões
            </h1>
            <Button
              active={false}
              disabled={false}
              onClick={[Function]}
              variant="primary"
            >
              <button
                className="btn btn-primary"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Log Out
              </button>
            </Button>
          </div>
          <hr />
          <FormGroup>
            <div
              className="form-group"
            >
              <FormControl
                id="email"
                name="email"
                onChange={[Function]}
                placeholder="Buscar pelo título"
                type="email"
              >
                <input
                  className="form-control"
                  id="email"
                  name="email"
                  onChange={[Function]}
                  placeholder="Buscar pelo título"
                  type="email"
                />
              </FormControl>
            </div>
          </FormGroup>
          <hr />
          <div
            style={
              Object {
                "marginBottom": "15px",
                "textAlign": "center",
              }
            }
          >
            <Button
              active={false}
              disabled={false}
              onClick={[Function]}
              variant="primary"
            >
              <button
                className="btn btn-primary"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Nova opinião
              </button>
            </Button>
          </div>
          <Card
            body={false}
            key="0"
            style={
              Object {
                "marginBottom": "15px",
              }
            }
          >
            <div
              className="card"
              style={
                Object {
                  "marginBottom": "15px",
                }
              }
            >
              <CardBody>
                <div
                  className="card-body"
                >
                  <CardTitle
                    style={
                      Object {
                        "alignItems": "center",
                        "display": "flex",
                        "justifyContent": "space-between",
                      }
                    }
                  >
                    <ForwardRef
                      className="card-title"
                      style={
                        Object {
                          "alignItems": "center",
                          "display": "flex",
                          "justifyContent": "space-between",
                        }
                      }
                    >
                      <div
                        className="card-title h5"
                        style={
                          Object {
                            "alignItems": "center",
                            "display": "flex",
                            "justifyContent": "space-between",
                          }
                        }
                      >
                        <div>
                          test_title
                        </div>
                        <div
                          style={
                            Object {
                              "color": "#28a745",
                              "fontWeight": "bold",
                            }
                          }
                        >
                          (
                          3
                          ) Votos
                        </div>
                      </div>
                    </ForwardRef>
                  </CardTitle>
                  <CardText>
                    <p
                      className="card-text"
                    >
                      <div
                        className="opinion-description"
                      >
                        <div
                          className="shadow"
                        />
                        <ReactMarkdown
                          astPlugins={Array []}
                          escapeHtml={true}
                          parserOptions={Object {}}
                          plugins={Array []}
                          rawSourcePos={false}
                          renderers={Object {}}
                          skipHtml={false}
                          source="test_content"
                          sourcePos={false}
                          transformLinkUri={[Function]}
                        >
                          <Root
                            key="root-1-1-0"
                          >
                            <p
                              key="paragraph-1-1-0"
                            >
                              <TextRenderer
                                key="text-1-1-0"
                                nodeKey="text-1-1-0"
                                value="test_content"
                              >
                                test_content
                              </TextRenderer>
                            </p>
                          </Root>
                        </ReactMarkdown>
                      </div>
                      <Button
                        active={false}
                        disabled={false}
                        onClick={[Function]}
                        style={
                          Object {
                            "marginTop": "10px",
                          }
                        }
                        variant="link"
                      >
                        <button
                          className="btn btn-link"
                          disabled={false}
                          onClick={[Function]}
                          style={
                            Object {
                              "marginTop": "10px",
                            }
                          }
                          type="button"
                        >
                          Exibir Detalhes
                        </button>
                      </Button>
                    </p>
                  </CardText>
                </div>
              </CardBody>
            </div>
          </Card>
        </div>
      </Container>
      <CreateModal
        onClose={[Function]}
        visible={false}
      >
        <Modal
          animation={true}
          autoFocus={true}
          backdrop={true}
          dialogAs={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "ModalDialog",
              "render": [Function],
            }
          }
          enforceFocus={true}
          keyboard={true}
          onHide={[Function]}
          restoreFocus={true}
          show={false}
        >
          <Modal
            autoFocus={true}
            backdrop={true}
            backdropTransition={[Function]}
            containerClassName="modal-open"
            enforceFocus={true}
            keyboard={true}
            manager={
              BootstrapModalManager {
                "containers": Array [],
                "data": Array [],
                "handleContainerOverflow": true,
                "hideSiblingNodes": true,
                "modals": Array [],
                "scrollbarSize": 0,
              }
            }
            onEnter={[Function]}
            onEntering={[Function]}
            onEscapeKeyDown={[Function]}
            onExit={[Function]}
            onExited={[Function]}
            onHide={[Function]}
            renderBackdrop={[Function]}
            renderDialog={[Function]}
            restoreFocus={true}
            show={false}
            transition={[Function]}
          />
        </Modal>
      </CreateModal>
      <DetailsModal
        onClose={[Function]}
        visible={false}
      >
        <Modal
          animation={true}
          autoFocus={true}
          backdrop={true}
          dialogAs={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "ModalDialog",
              "render": [Function],
            }
          }
          enforceFocus={true}
          keyboard={true}
          onHide={[Function]}
          restoreFocus={true}
          show={false}
        >
          <Modal
            autoFocus={true}
            backdrop={true}
            backdropTransition={[Function]}
            containerClassName="modal-open"
            enforceFocus={true}
            keyboard={true}
            manager={
              BootstrapModalManager {
                "containers": Array [],
                "data": Array [],
                "handleContainerOverflow": true,
                "hideSiblingNodes": true,
                "modals": Array [],
                "scrollbarSize": 0,
              }
            }
            onEnter={[Function]}
            onEntering={[Function]}
            onEscapeKeyDown={[Function]}
            onExit={[Function]}
            onExited={[Function]}
            onHide={[Function]}
            renderBackdrop={[Function]}
            renderDialog={[Function]}
            restoreFocus={true}
            show={false}
            transition={[Function]}
          />
        </Modal>
      </DetailsModal>
    </div>
  </Feed>
</Provider>
`;
